### PR TITLE
Creating Word Clouds in Python

### PR DESCRIPTION
**Pull request by lalith1403**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      

#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tool:** [Grammarly](https://grammarly.com)
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by lalith1403** [comment]
this fixes #612

**Comment by hectorkambow** [comment]
Plagiarism check done

**Comment by adrianmurage** [comment]
Hi @lalith1403 
I'll be conducting the review in two stages:
- Preliminary review where I'll look out for: grammar, spelling, and readability
- Tech Review where I'll look out for: technical accuracy, beginner-friendliness, and currency

**Comment by adrianmurage** [comment]
@lalith1403 
You have done a comprehensive job with the grammar part of this article.
I'll be moving on to the tech review

File: articles/word-cloud/index.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        data = [item.lower() for item in data]
+        # load the stop_words of english
+        stop_words = set(stopwords.words('english'))
+        # concatenate all the data with spaces.
+        para = ' '.join(data)
+        # tokenize the paragraph using the inbuilt tokenizer
+        word_tokens = word_tokenize(para) 
+        # filter words present in stopwords list 
+        preprocessed_data = ' '.join([w for w in word_tokens if not w in stop_words])
+        print("\n Preprocessed Data": ,preprocessed_data)
```
Comment (adrianmurage) [inline articles/word-cloud/index.md:56]: syntax error on this line

**Review by adrianmurage — CHANGES_REQUESTED** [review]
While trying to run this, I ran into a few problems:

- I had to create a virtual env to isolate the installations
- I had to install nltk which was not mentioned in the guide
- even after installing nltk I ran into an error that said I need to download stopwords

I advise that you add a virtual env setup to ensure the people who follow this guide all have the same environments.
Below is the error I am running into:
```bash
Traceback (most recent call last):
  File "/home/murage/wordcloud/virtual/lib/python3.6/site-packages/nltk/corpus/util.py", line 83, in __load
    root = nltk.data.find("{}/{}".format(self.subdir, zip_name))
  File "/home/murage/wordcloud/virtual/lib/python3.6/site-packages/nltk/data.py", line 585, in find
    raise LookupError(resource_not_found)
LookupError: 
**********************************************************************
  Resource stopwords not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('stopwords')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load corpora/stopwords.zip/stopwords/

  Searched in:
    - '/home/murage/nltk_data'
    - '/home/murage/wordcloud/virtual/nltk_data'
    - '/home/murage/wordcloud/virtual/share/nltk_data'
    - '/home/murage/wordcloud/virtual/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/murage/wordcloud/pyth.py", line 37, in <module>
    clean_data = wordcloudgen.preprocessing(data)
  File "/home/murage/wordcloud/pyth.py", line 15, in preprocessing
    stop_words = set(stopwords.words('english'))
  File "/home/murage/wordcloud/virtual/lib/python3.6/site-packages/nltk/corpus/util.py", line 120, in __getattr__
    self.__load()
  File "/home/murage/wordcloud/virtual/lib/python3.6/site-packages/nltk/corpus/util.py", line 85, in __load
    raise e
  File "/home/murage/wordcloud/virtual/lib/python3.6/site-packages/nltk/corpus/util.py", line 80, in __load
    root = nltk.data.find("{}/{}".format(self.subdir, self.__name))
  File "/home/murage/wordcloud/virtual/lib/python3.6/site-packages/nltk/data.py", line 585, in find
    raise LookupError(resource_not_found)
LookupError: 
**********************************************************************
  Resource stopwords not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('stopwords')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load corpora/stopwords

  Searched in:
    - '/home/murage/nltk_data'
    - '/home/murage/wordcloud/virtual/nltk_data'
    - '/home/murage/wordcloud/virtual/share/nltk_data'
    - '/home/murage/wordcloud/virtual/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************
```

My advice is to create a virtual env, try to follow through with the article, note the steps then make the necessary adjustments to the article.

**Comment by lalith1403** [comment]
Good catch @adrianmurage . I totally forgot mentioning the installation part! Will do so! Yes... Nltk installation is a big step! Additional download is required to get the stopwords. Apart from that could you execute the code? Also... Better thing would be to insert a repl.it link. @hectorkambow the installation varies from system to system and it becomes tedious to explain the installation every time. Can I include a repl link?

**Comment by lalith1403** [comment]
Including a repl link by passes the installation and makes it easier for the reader to follow. @hectorkambow let me know what you think of this.

**Comment by hectorkambow** [comment]
🤔  
Hmmm good ? - we can add repli.it links as long as the link the reader to the repl.it website - lets be sure not to include iframes. 
Can we do that? @lalith1403

**Comment by lalith1403** [comment]
That way can we skip the detailed installation section?

**Comment by hectorkambow** [comment]
> Good catch @adrianmurage . I totally forgot mentioning the installation part! Will do so! Yes... Nltk installation is a big step! Additional download is required to get the stopwords. Apart from that could you execute the code? Also... Better thing would be to insert a repl.it link. @hectorkambow the installation varies from system to system and it becomes tedious to explain the installation every time. Can I include a repl link?

Or could we include a sentence with a tutorial(s) to install on the 3 common systesm iOS linux and Windows. So we dont have to write out the whole installation process - just direct them to tutorials that walk through that part - so they can follow along ? 🤔 
@lalith1403

**Comment by lalith1403** [comment]
Since this is an application, I am worried if installation of nltk will take the limelight and not it's application.

**Comment by lalith1403** [comment]
@adrianmurage we decided to include a gentle introduction to installation of the required libraries as well. I will do so over the weekend. Thank you.

**Comment by adrianmurage** [comment]
@lalith1403 
Awesome
Ping me when you've made the changes and I'll take a look and give my feedback

**Comment by lalith1403** [comment]
@adrianmurage I have made the requested changes. Kindly take a look whenever you get a chance.

**Comment by lalith1403** [comment]
I think the front matter is needed. Can you kindly check that as well @adrianmurage @hectorkambow? Thank you.

File: articles/word-cloud/index.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+We define a class called `WordCloudGeneration` and define the following methods in the class:
+1. ** pre-processing**: We pass the input `data` through the tokenizer. The `data` is converted to lower case and tokenized. Tokenization results in a list of words. This list of words is further filtered. The filtering process copies words to `preprocessed_data` only if the word is not a stopword.
+
+2. **create_word_cloud**: This function takes in the processed list of words and calls the `WordCloud` class object. The `generate` method in the `WordCloud` class returns an image of the word cloud. Using the library `matplotlib`, we plot the image. 
+
+
+```py
+import matplotlib.pyplot as plt
+
+from nltk.corpus import stopwords
```
Comment (adrianmurage) [inline articles/word-cloud/index.md:72]: To be able to use this line the reader should have downloaded the `stopwords` and `punkt` packages using the [nltk downloader](https://www.nltk.org/data.html)
You should add that step to the installation.

File: articles/word-cloud/index.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        # convert all words to lowercase
+        data = [item.lower() for item in data]
+        # load the stop_words of english
+        stop_words = set(stopwords.words('english'))
+        # concatenate all the data with spaces.
+        para = ' '.join(data)
+        # tokenize the paragraph using the inbuilt tokenizer
+        word_tokens = word_tokenize(para) 
+        # filter words present in stopwords list 
+        preprocessed_data = ' '.join([w for w in word_tokens if not w in stop_words])
```
[the remark this answers is not in the document]
Comment (adrianmurage) [reply articles/word-cloud/index.md:87]: Kindly make the above-requested change

**Review by adrianmurage — COMMENTED** [review]
There isn't a clear way of seeing the word cloud image pointed out in the article.
I'm assuming one would need to be working in a Jupyter Notebook environment for them to see the image.

If this is the case. Point it out and guide them to a resource that guides them on how to do that.

File: articles/word-cloud/index.md (hunk 2 above)
Comment (lalith1403) [reply articles/word-cloud/index.md:72]: I created a new setup for this tutorial. There was no such requirement on my end. Can you tell me what the error is.

**Comment by lalith1403** [comment]
No! A jupyter notebook isn't needed. It pops up in another window. A matplotlib window.

**Review by adrianmurage — APPROVED** [review]
@lalith1403 
Great job with this

@hectorkambow 
This article is ready for final review

**Pull request merged**